### ISSUE TITLE
RequestEvaluator.GetRequestIsAppropriateForCasAuthentication() seems incomplete and appears to be causing redirect loop

### ISSUE THREAD

**Issue by rdev5** [Jasig/dotnet-cas-client#25]
We've been tracing a redirect loop issue that appears to be stemming from a seemingly incomplete implementation of `GetRequestIsAppropriateForCasAuthentication()`

Given the following (and only) route in RouteConfig.cs:

````
            routes.MapRoute(
                name: "Default",
                url: "{controller}/{action}",
                defaults: new { controller = "Login", action = "Index" }
            );
````

Browsing our SSO application (complete rewrite of CAS-OWA) at https://example.com/app/ works just fine and loads the LoginController#Index method.
````
200: https://example.com/app/
302: https://example.com/app/Login/SSO
302: https://cas.example.com/login?service=https://example.com/app/Login/SSO
302: https://example.com/app/Login/SSO?ticket=...
302: https://example.com/app/Login/SSO
200: https://example.com/finalDestination
...
````

However, when browsing to the following, we get a much different behavior:
````
200: https://example.com/app (note the missing trailing slash)
302: https://cas.example.com/login?service=https://example.com/app
302: https://example.com/app?ticket=...
302: https://example.com/app
302: https://cas.example.com/login?service=https://example.com/app
302: https://example.com/app?ticket=...
302: https://example.com/app
...
````

What is happening here is CasAuthenticationModule.cs specifies an `OnAuthenticateRequest` event handler being fired on every HttpRequest. I noticed that NEC-55 did not get address/merged so I did this myself in our local copy of the project:

````
        private static void OnAuthenticateRequest(object sender, EventArgs e)
        {
            // Validate the ticket coming back from the CAS server
            if (!RequestEvaluator.GetRequestIsAppropriateForCasAuthentication())
            {
                logger.Debug("AuthenticateRequest bypassed for " + HttpContext.Current.Request.RawUrl);
                return;
            }

            // TEST: Prevent processing at /app (caused by route)
            if (UrlUtil.RemoveCasArtifactsFromUrl(HttpContext.Current.Request.Url.AbsoluteUri) == "https://example.com/app")
            {
                HttpContext.Current.Response.Write("This request should NOT be getting authenticated");
                HttpContext.Current.Response.End();
                return;
            }

            HttpContext context = HttpContext.Current;
            HttpRequest request = context.Request;

            // Validate the ticket coming back from the CAS server
            if (RequestEvaluator.GetRequestHasCasTicket())
            {
                logger.Info("Processing Proxy Callback request");
                CasAuthentication.ProcessTicketValidation();
            }

            logger.Debug("Starting AuthenticateRequest for " + request.RawUrl);
            CasAuthentication.ProcessRequestAuthentication();
            logger.Debug("Ending AuthenticateRequest for " + request.RawUrl);
        }
````

In the latter example (https://example.com/app), I don't feel that the RequestEvaluator#GetRequestIsAppropriateForCasAuthentication is completely evaluating if the current request is in fact appropriate for CasAuthentication because under these conditions (using the /app path which .NET is not getting a chance to redirect to /app/ due to the module's current behavior), DotNetCasClient deems both of the following controller methods to be "appropriate for CasAuthentication:"

````
public class LoginController : Controller
   ...
   // NOT appropriate for CasAuthentication but still being processed
   // Not marked as "appropriate" when browsing /app/ (correct)
   // Marked as "appropriate" when browsing /app (incorrect)
   public void Index() {}

   // Appropriate for CasAuthentication
   [Authorize]
   public void SSO()
````

It appears that a quick remedy is to specify bypassCasForHandlers in the casClientConfig section which I can do, but it should also be noted that such a configuration option is not documented here: https://wiki.jasig.org/display/casc/.net+cas+client

Going back to the original method in question (GetRequestIsAppropriateForCasAuthentication), I'm assuming there isn't a practical way to evaluate if the current requested class/method has an [Authorize] attribute, or is there?

**Comment by rdev5** [comment]
Eh, one more thing to note which could be looked at either way (e.g. should have specified vs. shouldn't have mattered)...

I had `serverName` in casClientConfig specified as https://example.com/app instead of https://example.com/app/ and am going to test this now to see if merely adding the trailing slash resolves the login misdirection...

**Comment by rdev5** [comment]
So after tracing my way back out to the client application, I found the cause for the redirect loop to actually be in the SSO web app itself when doing this in the Index method:

````
FormsAuthentication.SignOut();
// FormsAuthentication.RedirectToLoginPage(); // Bad: Redirect loop...
Response.Redirect(SSOSettings.SSOWebApp.ssoUrl); // Good: No redirect loop...
````

It appears I got notAuthorizedUrl and casServerLoginUrl mixed when thinking about what RedirectToLoginPage() was actually redirecting to. However, it did help me to discover the uncommitted NETC-55 issue which may have easily become the next problem if it wasn't for this rabbit trail. I'll be closing this issue, pending the submission of that pull request.

Thanks again
